src/Stack/Runners.hs

```diff
@@ -136,10 +138,27 @@ withBuildConfigAndLockNoDocker
     -> (Maybe FileLock -> RIO EnvConfig ())
     -> IO ()
 withBuildConfigAndLockNoDocker go inner =
-    withBuildConfigExt True go Nothing inner Nothing
+    withBuildConfigExt True False go Nothing inner Nothing
+
+withBuildConfigAndLockInClean
+    :: GlobalOpts
+    -> (Maybe FileLock -> RIO EnvConfig ())
+    -> IO ()
+withBuildConfigAndLockInClean go inner =
+    withBuildConfigExt False True go Nothing inner Nothing
+
+-- | See issue #2010 for why this exists. Currently just used for the
+-- specific case of "stack clean --full".
+withBuildConfigAndLockNoDockerInClean
+    :: GlobalOpts
+    -> (Maybe FileLock -> RIO EnvConfig ())
+    -> IO ()
+withBuildConfigAndLockNoDockerInClean go inner =
+    withBuildConfigExt True True go Nothing inner Nothing
 
 withBuildConfigExt
     :: Bool
+    -> Bool -- ^ If perform the stack clean
```

Comment:
I'm concerned by the boolean blindness, here. The following types would be helpful.

```haskell
data WithDocker
  = SkipDocker
  | WithDocker

data IsClean
  = NotClean
  | IsClean
```

But see my comments on whether the docker checks are still necessary.

Reply:
Agreed, making another type that's iso to Bool is usually better.

To me the naming seems too purpose specific.  I think it makes sense to instead have the name correspond to the behavioral change.  That, is the name should specify whether or not it requires ghc to be installed.